feat: DoIP (iso 13400) client support

Magolves opened the pull request:
## Add DoIP (ISO 13400) Transport Layer Support

This PR implements DoIP (Diagnostic over IP - ISO 13400) as a new transport layer for UDS, enabling diagnostic communication over TCP/IP networks.

### Features

- **DoIP Client Implementation** (`src/tp/doip/`)
  - TCP-based diagnostic message transport (ISO 13400-2)
  - Routing activation and alive check handling
  - Proper DoIP message framing and state management
  - Support for diagnostic message ACK/NACK
  - Non-blocking I/O with configurable timeouts

### Implementation Details

- **Protocol Version**: DoIP v3 (ISO 13400:2019)
- **Standard Port**: TCP 13400
- **Addressing**: External test equipment range (0x0E00-0x0FFF)
- **Message Types**: Routing activation, diagnostic messages, alive check

### Testing

- Compiles cleanly with `-Wall -Wpedantic -Wextra`
- Integration test passes (`examples/doip_client/test.sh`)
- Successfully exchanges RDBI/WDBI messages over DoIP (WDBI reports NRC -> intended)

### Code Quality

- Consistent with iso14229 coding style
- Comprehensive error handling and logging
- Well-documented with inline comments and function documentation
- No compiler warnings

### Missing Features

- UDP vehicle discovery
- DoIP server implementation in `src/tp/doip/` (currently only in examples)
- TLS support (port 3496)
- Unit tests for DoIP module
- Multi-client server support

### Related Standards

- ISO 13400-2:2019 - Road vehicles — Diagnostic communication over Internet Protocol (DoIP)
- ISO 14229 - Unified Diagnostic Services (UDS)

---

**Testing Instructions:**

```bash
cd examples/doip_client
make
[test.sh](http://_vscodecontentref_/3)  # Automated integration test

Magolves committed:
feat: Add DoIP definitions

chore: Add CMakeList.txt

doc: Add notes on setting up build env

chore: Add option UDS_TP_DOIP

feat: Add DoIP header

chore: Add source files; integrate files into amalgate.py

fix: Begin impl of DoIP transport

fix: Pass tp pointer to all functions, remove global instance

feat: Add test code

fix: Update target address in DoIP client initialization

chore: Ignore example binaries

WIP

WIP: Corrected doip_poll/recv

feat: Add DOIP_DIAG_HEADER_SIZE

fix: Align return values of send/recv/poll to expected values (not finished)

test: Use sequence  from wdbi/rdbi test

test: Use verbose level

chore: Update amalgated code

doc: describe state machine

Magolves committed:
feat: Add DoIP definitions

Magolves committed:
chore: Add CMakeList.txt

Magolves committed:
doc: Add notes on setting up build env

Magolves committed:
chore: Add option UDS_TP_DOIP

Magolves committed:
feat: Add DoIP header

Magolves committed:
chore: Add source files; integrate files into amalgate.py

Magolves committed:
fix: Begin impl of DoIP transport

Magolves committed:
fix: Pass tp pointer to all functions, remove global instance

Magolves committed:
feat: Add test code

Magolves committed:
fix: Update target address in DoIP client initialization

Magolves committed:
chore: Ignore example binaries

Magolves committed:
WIP

Magolves committed:
feat: Enhance DoIP client state handling and update function return types

doc: Add poll states

fix: Fix doip_tp_poll  to return proper status

fix: Add doip_store_uds_response

refactor: Simplify DoIP address extraction with macro

test: Rename source files, update Makefile

fix: Fix DOIP_ADDRESS macro

test: Add test script for doip test

style: Rename global vars

fix: Respond properly to keep-alive requests

test: Add option to poll forever to test keep-alive

Magolves committed:
style: Code cleanup, add missing doc

fix: Fix compiler warnings
style: Code cleanup, adjust style, add README
doc: Add comments and refs to iso13400
doc: Add DoIP client to mainpage
chore: Add DoIP client test to CI

Magolves committed:
fix: Add safety belt for strncpy (sonar finding) and test case for invalid IP

Magolves committed:
test: Handle also WDBI messages

Magolves committed:
fix: Use snprintf instead of strncpy to make SonarCube happy

Magolves committed:
fix: Add missing const/sonarcube note

Magolves committed:
fix: Manual merge of diverted branches

Magolves committed:
Merge branch 'main' of https://github.com/Magolves/iso14229

driftregion commented:
Thanks @Magolves for this PR!

I have not given this PR the full review it deserves but a few things stick out:
- `__attribute__((packed))` is not cross-platform and must be removed.
- I see this targets unix. That is a great target, but it complicates testing (unit, fuzz) when the transport logic is coupled to the OS API.
- Vehicle discovery is required by the standard (https://www.iso.org/standard/74785.html). It seems like this would be a good place to start testing.

Do you intend to merge this PR or is this a demo?

Magolves commented:
> Thanks @Magolves for this PR!

Thanks for the hints - I will follow up on this

> Do you intend to merge this PR or is this a demo?

My intention was to merge it some day, yes.

Magolves committed:
fix: Remove __attribute(packed) (review comment)

Magolves committed:
feat: Begin TP abstraction (tcp and udp)

Magolves committed:
feat: Add DoIP UDP transport discovery and selection callback support

Magolves committed:
feat: Implement DoIP discovery example with VIN selection callback

Magolves committed:
doc: Update ReadMe

Magolves committed:
feat: Add DoIP discovery example with VIN selection callback to README

Magolves committed:
feat: Add Makefile

driftregion commented:
> My intention was to merge it some day, yes.

I am glad to hear. If a DoIP module is to be included in iso14229 in the future, these are my expectations:

- portable: the .c/.h modules which implement DoIP business logic (discovery, timeouts and other behavior described in the standard) must be portable, containing no os-specific code and making minimal assumptions about the host system.
- non-blocking: iso14229 is non-blocking by design and it expects the transports to be non-blocking also. For more details please see #38 and https://github.com/driftregion/iso14229/blob/main/src/tp/isotp_sock.c#L81. 
- static memory allocation
- testable: key standard areas are unit tested and branch coverage is filled out by fuzz testing (see: https://github.com/driftregion/iso14229/tree/main/fuzz ).

Please consider whether these goals are consistent with your own. Thanks!

Magolves committed:
fix: Make sockets non-blocking, define timeout via config, handle MSG_NOSIGNAL

Magolves committed:
fix: Fix & improve UDP discovery
chore: Fix build of discovery example

Magolves committed:
doc: Add ReadMe on DoIP client

Magolves committed:
WIP: Add marker for TP layer

Magolves committed:
Merge branch 'main' of https://github.com/Magolves/iso14229

Magolves committed:
Refactor DoIP transport layer to support mock implementations

- Updated DoIPClient to use function pointers for TCP and UDP transport methods, allowing for easier testing with mock implementations.
- Introduced DoIPTcpTransport and DoIPUdpTransport structures to encapsulate transport-specific data and methods.
- Replaced direct calls to transport functions with calls through the new function pointers.
- Implemented mock transport functions for both TCP and UDP to facilitate unit testing without actual network dependencies.
- Enhanced error handling and logging for transport initialization and operations.
- Updated related functions and headers to reflect the new transport structure and function pointer usage.

Magolves committed:
fix: Type, time overflow

Magolves committed:
fix: Use shared RX buffer, relief stack
